Services/UI/OJS.Services.Ui.Business/Implementations/ContestCategoriesBusinessService.cs

```diff
@@ -0,0 +1,73 @@
+namespace OJS.Services.Ui.Business.Implementations;
+
+using FluentExtensions.Extensions;
+using Microsoft.EntityFrameworkCore;
+using OJS.Services.Common.Models.Cache;
+using OJS.Services.Ui.Data;
+using SoftUni.AutoMapper.Infrastructure.Extensions;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+public class ContestCategoriesBusinessService : IContestCategoriesBusinessService
+{
+    private readonly IContestCategoriesDataService contestCategoriesData;
+
+    public ContestCategoriesBusinessService(
+        IContestCategoriesDataService contestCategoriesData)
+        => this.contestCategoriesData = contestCategoriesData;
+
+    public async Task<IEnumerable<ContestCategoryTreeViewModel>> GetTree()
+    {
+        var categories = await this.contestCategoriesData
+            .GetAllVisibleMainOrdered<ContestCategoryTreeViewModel>()
+            .ToListAsync();
+
+        await categories.ForEachSequential(async category =>
+            await AddChildren(category.Children));
```

Comment:
This is `N+1`. For 15 categories, this will make `1+15` calls to the DB.

Better to get all categories in-memory, then group.